kvredis/src/main.rs

```diff
@@ -0,0 +1,307 @@
+        &self,
+        ctx: &Context,
+        arg: &StringList,
+    ) -> Result<StringList, RpcError> {
+        let mut cmd = redis::Cmd::sinter(arg);
+        let value: Vec<String> = self.exec(ctx, &mut cmd).await?;
+        Ok(value)
+    }
+
+    async fn set_query<TS: ToString + ?Sized + Sync>(
+        &self,
+        ctx: &Context,
+        arg: &TS,
+    ) -> RpcResult<StringList> {
+        let mut cmd = redis::Cmd::smembers(arg.to_string());
+        let values: Vec<String> = self.exec(ctx, &mut cmd).await?;
+        Ok(values)
+    }
+
+    async fn set_union(&self, ctx: &Context, arg: &StringList) -> RpcResult<StringList> {
+        let mut cmd = redis::Cmd::sunion(arg);
+        let values: Vec<String> = self.exec(ctx, &mut cmd).await?;
+        Ok(values)
+    }
+}
+
+impl KvRedisProvider {
+    /// Helper function to execute redis async command while holding onto a mutable connection.
+    ///
+    /// This provider is multi-threaded, and requests from different actors use
+    /// different connections, and requests can run in parallel.
+    ///
+    /// There is a single connection per actor public key, and the write lock on the connection
+    /// effectively serializes redis operations for all instances of the same actor.
+    /// The lock is held only for the duration of a redis command from this provvider
+    /// and waiting for its response. The lock duration does not overlap with
+    /// message passing between actors and this provider, including serialization
+    /// of requests and deserialization of responses, which are fully parallelizable.
+    ///
+    /// There is a read lock held on the actors hashtable, which does not interfere
+    /// with redis operations, but any control commands for new actor links
+    /// or removal of actor links may need to wait for in-progress operations to complete.
+    /// That should be rare, because most links are passed to the provider at startup.
+    async fn exec<T: FromRedisValue>(&self, ctx: &Context, cmd: &mut redis::Cmd) -> RpcResult<T> {
+        let actor_id = ctx
+            .actor
+            .as_ref()
+            .ok_or_else(|| RpcError::InvalidParameter("no actor in request".to_string()))?;
+        // get read lock on actor-connections hashmap
+        let rd = self.actors.read().await;
+        let rc = rd
+            .get(actor_id)
+            .ok_or_else(|| RpcError::InvalidParameter(format!("actor not linked:{}", actor_id)))?;
+        // get write lock on this actor's connection
+        let mut con = rc.write().await;
+        cmd.query_async(con.deref_mut()).await.map_err(to_rpc_err)
+    }
+}
+
+/// Handle incoming rpc messages and dispatch them to applicable trait handler.
+#[async_trait]
+impl MessageDispatch for KvRedisProvider {
+    async fn dispatch(&self, ctx: &Context, message: Message<'_>) -> RpcResult<Message<'_>> {
+        let op = match message.method.split_once('.') {
+            Some((cls, op)) if cls == "KeyValue" => op,
```

Comment:
Forgive me if I missed this in the interface, does this mean all of our operations contain a `.` now? e.g.
- KeyValue.Set
- KeyValue.Add

Reply:
As long as the operation names are available as constants to actors and providers, no changes needed. Just looking for clarification

Reply:
the benefit is that no constants are required. OP constants don't need to be declared, and you avoid the bugs we had before of misspelling them and then wondering why the message wasn't received. The "client" sdk adds the method name in, and the dispatcher uses it.